libraries/AP_Baro/AP_Baro_ICM20789.cpp

```diff
@@ -0,0 +1,318 @@
+
+#include <AP_Math/AP_Math.h>
+#include <DataFlash/DataFlash.h>
+
+extern const AP_HAL::HAL &hal;
+
+/*
+  CMD_READ options. The draft datasheet doesn't specify, but it seems
+  Mode_1 has a conversion interval of 2ms. Mode_3 has a conversion
+  interval of 20ms. Both seem to produce equally as smooth results, so
+  presumably Mode_3 is doing internal averaging
+ */
+#define CMD_READ_PT_MODE_1 0x401A
+#define CMD_READ_PT_MODE_3 0x5059
+#define CMD_READ_TP_MODE_1 0x609C
+#define CMD_READ_TP_MODE_3 0x70DF
+
+#define CONVERSION_INTERVAL_MODE_1 2000
+#define CONVERSION_INTERVAL_MODE_3 20000
+
+// setup for Mode_3
+#define CMD_READ_PT CMD_READ_PT_MODE_3
+#define CONVERSION_INTERVAL CONVERSION_INTERVAL_MODE_3
+
+#define CMD_SOFT_RESET     0x805D
+#define CMD_READ_ID        0xEFC8
+
+#define debug(fmt, args...)   hal.console->printf(fmt, ##args)
+
+/*
+  constructor
+ */
+AP_Baro_ICM20789::AP_Baro_ICM20789(AP_Baro &baro, AP_HAL::OwnPtr<AP_HAL::Device> _dev)
+    : AP_Baro_Backend(baro)
+    , dev(std::move(_dev))
+{
+}
+
+AP_Baro_Backend *AP_Baro_ICM20789::probe(AP_Baro &baro,
+                                         AP_HAL::OwnPtr<AP_HAL::Device> dev)
+{
+    debug("Probing for ICM20789 baro\n");
+    if (!dev) {
+        return nullptr;
+    }
+    AP_Baro_ICM20789 *sensor = new AP_Baro_ICM20789(baro, std::move(dev));
+    if (!sensor || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+    return sensor;
+}
+
+
+bool AP_Baro_ICM20789::spi_init(void)
+{
+    dev_icm->set_read_flag(0x80);
+
+    dev_icm->set_speed(AP_HAL::Device::SPEED_LOW);
+    uint8_t whoami = 0;
+    uint8_t v;
+        
+    dev_icm->read_registers(0x6A, &v, 1);
+    dev_icm->write_register(0x6B, 0x01);
+    
+    hal.scheduler->delay(1);
+    dev_icm->write_register(0x6A, 0x10);
+    dev_icm->write_register(0x6B, 0x41);
+    
+    hal.scheduler->delay(1);
+    dev_icm->write_register(0x6B, 0x01);
+    
+    hal.scheduler->delay(1);
+    dev_icm->write_register(0x23, 0x00);
+    dev_icm->write_register(0x6B, 0x41);
+    
+    dev_icm->read_registers(0x75, &whoami, 1);
+    
+    // wait for sensor to settle
+    hal.scheduler->delay(100);
+    
+    dev_icm->read_registers(0x75, &whoami, 1);
+    
+    dev_icm->write_register(0x37, 0x00);
+    dev_icm->write_register(0x6A, 0x10);
+    return true;
+}
+
+bool AP_Baro_ICM20789::init()
+{
+    if (!dev) {
+        return false;
+    }
+
+    debug("Looking for 20789 baro\n");
+    
+    if (!dev->get_semaphore()->take(0)) {
+        AP_HAL::panic("PANIC: AP_Baro_ICM20789: failed to take serial semaphore for init");
+    }
+
+#ifdef HAL_INS_MPU60x0_NAME
+    /*
+      Pressure sensor data can be accessed in the following mode:
+      Bypass Mode: Set register INT_PIN_CFG (Address: 55 (Decimal); 37 (Hex)) bit 1 to value 1 
+      and I2C_MST_EN bit is 0 Address: 106 (Decimal); 6A (Hex)). 
+      Pressure sensor data can then be accessed using the procedure described in Section 10.
+    */
```

Comment:
This comment should be on the function above. And it looks strange... Are you setting it to read in bypass mode so you have both the  IMU on spi bus and baro on I2C bus (i.e. you have real access to both SPI pins and I2C pins and they are not the same)? This would be different from all the other Invensense sensors in which there's an I2C transaction over the SPI bus.  By the name of this register (I2C_MST_EN) it looks like you are indeed enabling the I2C master engine on the sensor.